std/experimental/allocator/package.d

```diff
@@ -0,0 +1,1940 @@
+
+If $(D array.length < delta), does nothing and returns `false`. Otherwise,
+destroys the last $(D array.length - delta) elements in the array and then
+reallocates the array's buffer. If reallocation fails, fills the array with
+default-initialized data.
+
+Params:
+T = element type of the array being created
+alloc = the allocator used for getting memory
+array = a reference to the array being shrunk
+delta = number of elements to remove (upon success the new length of $(D array) is $(D array.length - delta))
+
+Returns:
+`true` upon success, `false` if memory could not be reallocated. In the latter
+case, the slice $(D array[$ - delta .. $]) is left with default-initialized
+elements.
+
+Throws:
+The first two overloads throw only if `alloc`'s primitives do. The
+overloads that involve copy initialization deallocate memory and propagate the
+exception if the copy operation throws.
+*/
+bool shrinkArray(T, Allocator)(auto ref Allocator alloc,
+        ref T[] array, size_t delta)
+{
+    if (delta > array.length) return false;
+
+    // Destroy elements. If a destructor throws, fill the already destroyed
+    // stuff with the default initializer.
+    {
+        size_t destroyed;
+        scope(failure)
+        {
+            array[$ - delta .. $][0 .. destroyed].uninitializedFillDefault;
+        }
+        foreach (ref e; array[$ - delta .. $])
+        {
+            e.destroy;
+            ++destroyed;
+        }
+    }
+
+    if (delta == array.length)
+    {
+        alloc.deallocate(array);
+        array = null;
+        return true;
+    }
+
+    void[] buf = array;
+    if (!alloc.reallocate(buf, buf.length - T.sizeof * delta))
+    {
+        // urgh, at least fill back with default
+        array[$ - delta .. $].uninitializedFillDefault;
+        return false;
+    }
+    array = cast(T[]) buf;
+    return true;
+}
+
+///
+unittest
+{
+    int[] a = theAllocator.makeArray!int(100, 42);
+    assert(a.length == 100);
+    assert(theAllocator.shrinkArray(a, 98));
+    assert(a.length == 2);
+    assert(a == [42, 42]);
+}
+
+unittest
+{
+    void test(A)(auto ref A alloc)
+    {
+        long[] a = alloc.makeArray!long((int[]).init);
+        assert(a.length == 0 && a.ptr is null);
+        a = alloc.makeArray!long(100, 42);
+        assert(alloc.shrinkArray(a, 98));
+        assert(a.length == 2);
+        assert(a == [ 42, 42]);
+    }
+    import std.experimental.allocator.gc_allocator : GCAllocator;
+    test(GCAllocator.instance);
+    test(theAllocator);
+}
+
+/**
+
+Destroys and then deallocates (using $(D alloc)) the object pointed to by a
+pointer, the class object referred to by a $(D class) or $(D interface)
+reference, or an entire array. It is assumed the respective entities had been
+allocated with the same allocator.
+
+*/
+void dispose(A, T)(auto ref A alloc, T* p)
+{
+    static if (hasElaborateDestructor!T)
+    {
+        destroy(*p);
+    }
+    alloc.deallocate(p[0 .. T.sizeof]);
```

Comment:
This is wrong. It needs to be `alloc.deallocate((cast(void*) p)[0 .. T.sizeof]);`